File: frontend/js/views/mortgage_affordability.js
// Mortgage affordability — max home price under the 28/36 rule, via
// /calc/mortgage-affordability. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const VIEW = 'mortgage-affordability';
let lastReport = null;
let lastBody = null;

export async function renderMortgageAffordability(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.afford.h1.title">// MORTGAGE AFFORDABILITY</span></h1>
        <p class="muted small" data-i18n="view.afford.hint.intro">
            The most house you can buy under the 28/36 rule — lenders cap housing cost (PITI) at 28%
            of gross monthly income and total debt at 36%. The tighter cap sets the budget; since
            taxes and the loan payment both scale with price, it solves for the max home price.
            Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.afford.h2.inputs">Your finances</h2>
            <form id="afford-form" class="inline-form">
                <label><span data-i18n="view.afford.label.income">Annual income ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_income_usd" value="100000" required></label>
                <label><span data-i18n="view.afford.label.debts">Other monthly debts ($)</span>
                    <input type="number" step="0.01" min="0" name="monthly_debts_usd" value="500"></label>
                <label><span data-i18n="view.afford.label.down">Down payment ($)</span>
                    <input type="number" step="0.01" min="0" name="down_payment_usd" value="50000" required></label>
                <label><span data-i18n="view.afford.label.rate">Mortgage rate (%)</span>
                    <input type="number" step="0.001" min="0" name="annual_rate_pct" value="6.5" required></label>
                <label><span data-i18n="view.afford.label.term">Term (months)</span>
                    <input type="number" step="1" min="1" name="term_months" value="360" required></label>
                <label><span data-i18n="view.afford.label.tax">Property tax rate (% / yr)</span>
                    <input type="number" step="0.01" min="0" name="property_tax_rate_pct" value="1.2"></label>
                <label><span data-i18n="view.afford.label.insurance">Annual insurance ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_insurance_usd" value="1500"></label>
                <label><span data-i18n="view.afford.label.front">Front-end (%)</span>
                    <input type="number" step="0.1" min="0" name="front_end_pct" value="28"></label>
                <label><span data-i18n="view.afford.label.back">Back-end (%)</span>
                    <input type="number" step="0.1" min="0" name="back_end_pct" value="36"></label>
            </form>
            <div id="afford-tools" class="ce-toolbar"></div>
            <button type="button" id="afford-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="afford-sens" class="ce-sens"></div>
        </div>
        <div id="afford-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#afford-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            annual_income_usd: Number(fd.get('annual_income_usd')) || 0,
            monthly_debts_usd: Number(fd.get('monthly_debts_usd')) || 0,
            down_payment_usd: Number(fd.get('down_payment_usd')) || 0,
            annual_rate_pct: Number(fd.get('annual_rate_pct')) || 0,
            term_months: Number(fd.get('term_months')) || 0,
            property_tax_rate_pct: Number(fd.get('property_tax_rate_pct')) || 0,
            annual_insurance_usd: Number(fd.get('annual_insurance_usd')) || 0,
            front_end_pct: Number(fd.get('front_end_pct')) || 0,
            back_end_pct: Number(fd.get('back_end_pct')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcMortgageAffordability(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.afford.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#afford-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'mortgage-affordability.csv' });
    mount.querySelector('#afford-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['max_home_price_usd', r.max_home_price_usd],
        ['max_loan_usd', r.max_loan_usd],
        ['max_piti_usd', r.max_piti_usd],
        ['binding_constraint', r.binding_constraint],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#afford-result');
    const bindKey = r.binding_constraint === 'front' ? 'view.afford.bind.front' : 'view.afford.bind.back';
    // Line chart: max home price as the mortgage rate sweeps 3% -> 10% (affordability falls).
    const xs = enh.linspace(3, 10, 13);
    const pts = await Promise.all(xs.map(async (rate) => {
        const rr = await api.calcMortgageAffordability({ ...body, annual_rate_pct: rate });
        return { x: rate, y: rr ? rr.max_home_price_usd / 1000 : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'rate %', ylabel: 'max price $k' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.afford.h2.result">What you can afford</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.afford.card.price">Max home price</div>
                    <div class="value pos">${money(r.max_home_price_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.afford.card.loan">Max loan</div>
                    <div class="value">${money(r.max_loan_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.afford.card.piti">Max PITI / mo</div>
                    <div class="value">${money(r.max_piti_usd)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.afford.row.income">Monthly income</td><td>${money(r.monthly_income_usd)}</td></tr>
                    <tr><td data-i18n="view.afford.row.front">Front-end cap (28%)</td><td>${money(r.front_end_max_usd)}</td></tr>
                    <tr><td data-i18n="view.afford.row.back">Back-end room (36% − debts)</td><td>${money(r.back_end_max_usd)}</td></tr>
                    <tr><td data-i18n="view.afford.row.binding">Binding constraint</td><td data-i18n="${bindKey}">—</td></tr>
                    <tr><td data-i18n="view.afford.row.loan">Max loan</td><td>${money(r.max_loan_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.afford.row.price">Max home price</td><td>${money(r.max_home_price_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#afford-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: mortgage rate 3% -> 10%; y: annual income 0.6x -> 1.4x. Output: max home price.
    const inc = base.annual_income_usd || 100000;
    const xVals = enh.linspace(3, 10, 5);
    const yVals = enh.linspace(inc * 0.6, inc * 1.4, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'annual_rate_pct', yKey: 'annual_income_usd', xVals, yVals, compute: (b) => api.calcMortgageAffordability(b), pick: (r) => (r ? r.max_home_price_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells, fmt: (v) => (v == null ? '—' : '$' + Math.round(v / 1000) + 'k'), xfmt: (v) => v.toFixed(1) + '%', yfmt: (v) => '$' + Math.round(v / 1000) + 'k', xName: t('view.afford.label.rate') || 'Rate', yName: t('view.afford.label.income') || 'Income' });
}

File: crates/traderview-core/src/mortgage_affordability.rs
//! Mortgage affordability — the most house you can buy under the 28/36 rule.
//!
//! Lenders cap housing cost (PITI) at ~28% of gross monthly income (front-end)
//! and total debt at ~36% (back-end). The tighter cap sets the max PITI; since
//! property tax and the loan payment both scale with price, solve for it:
//!
//! ```text
//! max PITI = (price − down)·pf + price·(tax rate/12) + insurance/12
//! price = (max PITI − insurance/12 + down·pf) / (pf + tax rate/12)
//! ```
//!
//! where `pf` is the per-dollar mortgage payment factor.

use serde::{Deserialize, Serialize};

fn d_front() -> f64 {
    28.0
}
fn d_back() -> f64 {
    36.0
}

#[derive(Debug, Clone, Deserialize)]
pub struct AffordabilityInput {
    pub annual_income_usd: f64,
    /// Other monthly debt payments (cards, auto, student loans).
    #[serde(default)]
    pub monthly_debts_usd: f64,
    pub down_payment_usd: f64,
    pub annual_rate_pct: f64,
    pub term_months: f64,
    /// Annual property tax as a percent of home value.
    #[serde(default)]
    pub property_tax_rate_pct: f64,
    /// Annual homeowners insurance (dollars).
    #[serde(default)]
    pub annual_insurance_usd: f64,
    #[serde(default = "d_front")]
    pub front_end_pct: f64,
    #[serde(default = "d_back")]
    pub back_end_pct: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct AffordabilityResult {
    pub monthly_income_usd: f64,
    /// Front-end cap (housing only).
    pub front_end_max_usd: f64,
    /// Back-end cap less other debts (housing room within total-debt limit).
    pub back_end_max_usd: f64,
    /// The binding monthly housing budget (lesser of the two).
    pub max_piti_usd: f64,
    /// Largest loan the budget supports.
    pub max_loan_usd: f64,
    /// Largest home price (loan + down payment).
    pub max_home_price_usd: f64,
    /// Which cap binds: "front" or "back".
    pub binding_constraint: String,
}

pub fn analyze(input: &AffordabilityInput) -> AffordabilityResult {
    let monthly_income = input.annual_income_usd / 12.0;
    let front_max = input.front_end_pct / 100.0 * monthly_income;
    let back_max = (input.back_end_pct / 100.0 * monthly_income - input.monthly_debts_usd).max(0.0);

    let (max_piti, binding) = if front_max <= back_max {
        (front_max, "front")
    } else {
        (back_max, "back")
    };

    let i = input.annual_rate_pct / 100.0 / 12.0;
    let n = input.term_months;
    let pf = if i.abs() < 1e-12 {
        if n > 0.0 { 1.0 / n } else { 0.0 }
    } else {
        let f = (1.0 + i).powf(n);
        i * f / (f - 1.0)
    };

    let tax_monthly_factor = input.property_tax_rate_pct / 100.0 / 12.0;
    let insurance_monthly = input.annual_insurance_usd / 12.0;

    let denom = pf + tax_monthly_factor;
    let price = if denom > 0.0 {
        ((max_piti - insurance_monthly + input.down_payment_usd * pf) / denom).max(0.0)
    } else {
        0.0
    };
    let loan = (price - input.down_payment_usd).max(0.0);

    AffordabilityResult {
        monthly_income_usd: monthly_income,
        front_end_max_usd: front_max,
        back_end_max_usd: back_max,
        max_piti_usd: max_piti,
        max_loan_usd: loan,
        max_home_price_usd: price,
        binding_constraint: binding.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-2
    }

    fn base() -> AffordabilityInput {
        AffordabilityInput {
            annual_income_usd: 100_000.0,
            monthly_debts_usd: 500.0,
            down_payment_usd: 50_000.0,
            annual_rate_pct: 6.5,
            term_months: 360.0,
            property_tax_rate_pct: 1.2,
            annual_insurance_usd: 1_500.0,
            front_end_pct: 28.0,
            back_end_pct: 36.0,
        }
    }

    #[test]
    fn income_and_caps() {
        let r = analyze(&base());
        assert!(close(r.monthly_income_usd, 8_333.3333));
        assert!(close(r.front_end_max_usd, 2_333.3333));
        assert!(close(r.back_end_max_usd, 2_500.0));
    }

    #[test]
    fn front_end_binds_here() {
        let r = analyze(&base());
        assert_eq!(r.binding_constraint, "front");
        assert!(close(r.max_piti_usd, 2_333.3333));
    }

    #[test]
    fn max_price_and_loan() {
        let r = analyze(&base());
        assert!(close(r.max_home_price_usd, 344_826.8828));
        assert!(close(r.max_loan_usd, 294_826.8828));
    }

    #[test]
    fn back_end_binds_with_high_debts() {
        let r = analyze(&AffordabilityInput {
            monthly_debts_usd: 1_500.0,
            ..base()
        });
        // Back-end room = 3000 − 1500 = 1500 < front 2333 → back binds.
        assert_eq!(r.binding_constraint, "back");
        assert!(close(r.max_piti_usd, 1_500.0));
    }

    #[test]
    fn higher_income_affords_more() {
        let low = analyze(&base());
        let high = analyze(&AffordabilityInput {
            annual_income_usd: 200_000.0,
            ..base()
        });
        assert!(high.max_home_price_usd > low.max_home_price_usd);
    }

    #[test]
    fn larger_down_payment_raises_price() {
        let small = analyze(&base());
        let big = analyze(&AffordabilityInput {
            down_payment_usd: 100_000.0,
            ..base()
        });
        assert!(big.max_home_price_usd > small.max_home_price_usd);
    }

    #[test]
    fn higher_rate_lowers_price() {
        let low = analyze(&base());
        let high = analyze(&AffordabilityInput {
            annual_rate_pct: 9.0,
            ..base()
        });
        assert!(high.max_home_price_usd < low.max_home_price_usd);
    }

    #[test]
    fn loan_plus_down_equals_price() {
        let r = analyze(&base());
        assert!(close(r.max_loan_usd + 50_000.0, r.max_home_price_usd));
    }
}
